OPHD/StructureManager.cpp

```diff
@@ -452,6 +452,15 @@ void StructureManager::serialize(NAS2D::Xml::XmlElement* element)
 			structureElement->linkEndChild(food);
 		}
 
+		if (structure->structureClass() == Structure::StructureClass::Residence)
+		{
+			Residence* residence = static_cast<Residence*>(structure);
+			auto* waste = new NAS2D::Xml::XmlElement("waste");
+			waste->attribute("accumulated", residence->wasteAccumulated());
+			waste->attribute("overflow", residence->wasteOverflow());
+			structureElement->linkEndChild(waste);
+		}
```

Comment:
Looks like there are a lot of special case situations being handled in `StructureManager`. Maybe a `virtual` method should be added to `Structure` that returns a `Dictionary` of key/value pairs that need to be saved when writing a saved game file. The `StructureManager` can then translate the `Dictionary` to appropriate fields on an `XmlElement`.

At least that's assuming all such data that might be saved can be represented as key/value pairs.

You could also pass an `XmlElement` into a `virtual` method on `Structure` to have it write extra values. That would be more general, though it couples more of your code to the specifics of the XML library you're using. That may be worth avoiding since there is a plan to upgrade or replace that library.

Reply:
I like this idea though I don't have a complete idea on how to implement. If you get a chance opening a ticket detailing this change would be very helpful!